# Text Similarity - Natural Language Processing

## Introduction

Natural Language Processing (NLP) is an area of artificial intelligence that seeks to enable computers to understand, interpret, and produce human language efficiently.

## Objective

The objective of this project is to return a similarity score between sentences. For this project, the containment measure will be used, which is a measure of similarity between two texts.

## Data

Given the following original sentence:

> Olhando para a escala na parede, qual valor indicaria melhor a sua dor hoje?

We will compare the following 3 sentences, returning the similarity score between the original and the comparative.

|Original Sentence |Comparative Sentences | Similarity Score |
|--|--|--|
|Olhando para a escala na parede, qual valor indicaria melhor a sua dor hoje?|De acordo com a escala de dor ali na parede, qual valor você acha que mais representa a sua dor?|???
||De 0 a 10, qual o nível de intensidade da sua dor atualmente?|???
||Qual a intensidade da sua dor?|???


## Getting started

#### Importing Libraries

To start implementing the code, we will import some libraries that will be necessary for the project creation.

In [22]:
import numpy as np
import sklearn
import pandas as pd

#### Declaring Variables

We will declare some variables to store the original text and the comparative texts.

In [23]:
default_text = "Olhando para a escala na parede, qual valor indicaria melhor a sua dor hoje?"
text_1 = "De acordo com a escala de dor ali na parede, qual valor você acha que mais representa a sua dor?"
text_2 = "De 0 a 10, qual o nível de intensidade da sua dor atualmente?"
text_3 = "Qual a intensidade da sua dor?"

#### Creating Vocabulary

Now we will use Count Vectorizer to transform the words of the sentences that will be compared into a numeric format, to be used later in the comparisons. Remember that we will use unigrams, words with 2 or more letters will be used, and we will ignore single-letter words, as it is very common in text analysis.

In [24]:
from sklearn.feature_extraction.text import CountVectorizer

words = CountVectorizer(analyzer='word', ngram_range=(0,1))

texts = [text_1, text_2, text_3]

for text in texts:
    vocabulary = words.fit([text, default_text]).vocabulary_
    print(vocabulary)

{'de': 4, 'acordo': 1, 'com': 3, 'escala': 6, 'dor': 5, 'ali': 2, 'na': 11, 'parede': 14, 'qual': 15, 'valor': 19, 'você': 20, 'acha': 0, 'que': 16, 'mais': 9, 'representa': 17, 'sua': 18, 'olhando': 12, 'para': 13, 'indicaria': 8, 'melhor': 10, 'hoje': 7}
{'de': 3, '10': 0, 'qual': 15, 'nível': 11, 'intensidade': 8, 'da': 2, 'sua': 16, 'dor': 4, 'atualmente': 1, 'olhando': 12, 'para': 13, 'escala': 5, 'na': 10, 'parede': 14, 'valor': 17, 'indicaria': 7, 'melhor': 9, 'hoje': 6}
{'qual': 11, 'intensidade': 5, 'da': 0, 'sua': 12, 'dor': 1, 'olhando': 8, 'para': 9, 'escala': 2, 'na': 7, 'parede': 10, 'valor': 13, 'indicaria': 4, 'melhor': 6, 'hoje': 3}


#### Transforming Vocabularies into Arrays

Transforming the vocabularies into arrays for a comparison of each comparative sentence with the original sentence

In [25]:
for text in texts:
    array = words.fit_transform([text, default_text]).toarray()
    print(array)

[[1 1 1 1 2 2 1 0 0 1 0 1 0 0 1 1 1 1 1 1 1]
 [0 0 0 0 0 1 1 1 1 0 1 1 1 1 1 1 0 0 1 1 0]]
[[1 1 1 2 1 0 0 0 1 0 0 1 0 0 0 1 1 0]
 [0 0 0 0 1 1 1 1 0 1 1 0 1 1 1 1 1 1]]
[[1 1 0 0 0 1 0 0 0 0 0 1 1 0]
 [0 1 1 1 1 0 1 1 1 1 1 1 1 1]]


#### Creating a similarity measure

Now let's calculate the intersection between the comparative texts and the original text.

In [26]:
for text in texts:
    array = words.fit_transform([text, default_text]).toarray()
    intersections = np.sum(np.amin(array, axis=0))
    print(intersections)

7
3
3


#### Similarity Score

Given the previous comparison and calculating the similarity score, we obtain the following result.

In [27]:
divider = words.fit_transform([default_text]).toarray()
count_default = np.sum(divider)

for text in texts:
    array = words.fit_transform([text, default_text]).toarray()
    intersections = np.sum(np.amin(array, axis=0))
    score = intersections/count_default
    print(score)

0.5833333333333334
0.25
0.25


#### Creating the dataframe

For a better visualization, we will create a dataframe with the obtained results.

In [28]:
result_list = []

for text in texts:
    array = words.fit_transform([text, default_text]).toarray()
    intersections = np.sum(np.amin(array, axis=0))
    divider = words.fit_transform([default_text]).toarray()
    count_default = np.sum(divider)
    score = intersections/count_default
    result_dict = {'Original sentence': default_text, 'Comparative sentences': text, 'Similarity score': score}
    result_list.append(result_dict)

df = pd.DataFrame(result_list)

pd.set_option('display.max_rows', None)

df_output = pd.DataFrame(df, columns=['Original sentence', 'Comparative sentences', 'Similarity score'])
output_str = df_output.to_string(index=False)

print(output_str)


                                                           Original sentence                                                                            Comparative sentences  Similarity score
Olhando para a escala na parede, qual valor indicaria melhor a sua dor hoje? De acordo com a escala de dor ali na parede, qual valor você acha que mais representa a sua dor?          0.583333
Olhando para a escala na parede, qual valor indicaria melhor a sua dor hoje?                                    De 0 a 10, qual o nível de intensidade da sua dor atualmente?          0.250000
Olhando para a escala na parede, qual valor indicaria melhor a sua dor hoje?                                                                   Qual a intensidade da sua dor?          0.250000


#### Conclusion

Using the containment measure to calculate the similarity between two texts, we obtained the following result:

Original Phrase<br />
• Looking at the pain scale on the wall, what value would best indicate your pain today?<br />

Comparative Phrases<br />
• According to the pain scale on the wall, what value do you think best represents your pain?<br />
• On a scale of 0 to 10, what is the intensity level of your current pain?<br />
• What is the intensity of your pain?<br />

Similarity Scores<br />
• The first sentence obtained a score of 58.33%.<br />
• The second sentence obtained a score of 25%.<br />
• The third sentence obtained a score of 25%.<br />

Just as in this study, other models and measures could also be used, which would leave open to new improvements and consequently new results.